### ISSUE TITLE
AdamOptimizer.Optimize: initial PrepareAndEvaluateSolution does full-batch Train(XTrain, YTrain), OOMs Transformer at any non-trivial corpus size

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1296]
## Summary

For gradient-based optimizers (Adam / AdamW / SGD), `AdamOptimizer.Optimize` runs a redundant **full-batch** `Model.Train(XTrain, YTrain)` call before the mini-batched epoch loop begins. This blows past `AdamOptimizerOptions.BatchSize` (default 32) and OOMs on the very first forward pass whenever XTrain is large enough that one `MultiHeadAttention.Forward` blows the heap.

For a Transformer with `d=128, ff=512, heads=4, L=4, ctx=64` on a 1 MB byte corpus (~57500 training samples after stride=16), the first Train call allocates a `[57500, 4, 64, 64]` attention-scores tensor ≈ **3.8 GB float**, plus the LINQ `Vector<T>(IEnumerable<T>)` ctor inside `ScaledDotProductAttentionFloat` doubles it transiently → `OutOfMemoryException`.

This is independent of (and orthogonal to) the residual-leak fixes in #1227 / PR #1285 — those addressed cross-Train-call retention; this is a single-call OOM caused by skipping mini-batching on the initial evaluation pass.

## Repro

Minimum repro: any Transformer.Train via the `AiModelBuilder` facade on a large XTrain tensor.

```csharp
var arch = new TransformerArchitecture<float>(
    inputType: InputType.TwoDimensional,
    taskType: NeuralNetworkTaskType.SequenceClassification,
    numEncoderLayers: 4, numDecoderLayers: 0, numHeads: 4,
    modelDimension: 128, feedForwardDimension: 512,
    inputSize: 64, outputSize: 256,
    maxSequenceLength: 64, vocabularySize: 256);

var model = new Transformer<float>(arch, lossFunction: new CategoricalCrossEntropyLoss<float>());
var adamOpts = new AdamOptimizerOptions<float, Tensor<float>, Tensor<float>>
{
    InitialLearningRate = 3e-4,
    MaxIterations = 3,
    BatchSize = 32,           // <-- ignored on initial Train
    UseAdaptiveLearningRate = false,
};
var optimizer = new AdamOptimizer<float, Tensor<float>, Tensor<float>>(null, adamOpts);

// Build a batched [N, ctx] xTrain / [N, V] yTrain where N is large (~57500).
var loader = DataLoaders.FromTensors<float>(xTrain, yTrain);

var builder = new AiModelBuilder<float, Tensor<float>, Tensor<float>>(licenseKey)
    .ConfigureModel(model)
    .ConfigureOptimizer(optimizer)
    .ConfigureDataLoader(loader)
    .ConfigureGpuAcceleration(new GpuAccelerationConfig { UsageLevel = GpuUsageLevel.AlwaysCpu });

await builder.BuildAsync();   // OOMs here
```

Observed exception (relevant frames):

```
System.OutOfMemoryException
  at System.Linq.Enumerable.ICollectionToArray[TSource](ICollection`1 collection)
  at AiDotNet.Tensors.LinearAlgebra.VectorBase`1..ctor(IEnumerable`1 values)
  at AiDotNet.Tensors.Engines.CpuEngine.ScaledDotProductAttentionFloat(...)
  at AiDotNet.NeuralNetworks.Layers.MultiHeadAttentionLayer`1.Forward(Tensor`1 input)
  at AiDotNet.NeuralNetworks.NeuralNetworkBase`1.ForwardForTraining(Tensor`1 input)
  at AiDotNet.NeuralNetworks.NeuralNetworkBase`1.TrainWithTape(Tensor`1, Tensor`1, IGradientBasedOptimizer`3)
  at AiDotNet.NeuralNetworks.Transformer`1.Train(Tensor`1 input, Tensor`1 expectedOutput)
  at AiDotNet.Optimizers.OptimizerBase`3.TrainAndEvaluateSolution(ModelEvaluationInput`3, Boolean)
  at AiDotNet.Optimizers.OptimizerBase`3.PrepareAndEvaluateSolution(IFullModel`3, OptimizationInputData`3)
  at AiDotNet.Optimizers.AdamOptimizer`3.Optimize(OptimizationInputData`3 inputData)
  at AiDotNet.AiModelBuilder`3.BuildSupervisedInternalAsync(TInput x, TOutput y)
  at AiDotNet.AiModelBuilder`3.BuildAsync(CancellationToken)
```

`MaxIterations=3` never gets a chance to run — the OOM happens before the epoch loop starts. Runtime: ~5 min 12 sec on a 32 GB machine before the heap walks itself off the cliff.

## Root cause

`AdamOptimizer.Optimize` (`src/Optimizers/AdamOptimizer.cs:148`):

```csharp
var previousStepData = PrepareAndEvaluateSolution(currentSolution, inputData);   // <-- runs initial full-batch Train

for (int epoch = 0; epoch < _options.MaxIterations; epoch++)
{
    var batcher = CreateBatcher(inputData, _options.BatchSize);    // <-- correctly batched
    foreach (var (xBatch, yBatch, batchIndices) in batcher.GetBatches())
    {
        var gradient = CalculateGradient(currentSolution, xBatch, yBatch);
        currentSolution = UpdateSolution(currentSolution, gradient);
    }
    ...
}
```

`PrepareAndEvaluateSolution` (`OptimizerBase.cs:499`) calls `TrainAndEvaluateSolution(input, skipTraining: SkipTrainingInEvaluation)`. `SkipTrainingInEvaluation` is `false` initially (`GradientBasedOptimizerBase.cs:50`), so `TrainAndEvaluateSolution` (`OptimizerBase.cs:540`) executes:

```csharp
input.Model?.Train(input.InputData.XTrain, input.InputData.YTrain);   // <-- full-batch
```

i.e. the very first thing every gradient-based `Optimize` call does is push the **entire training tensor** through `Transformer.Train` -> `TrainWithTape` -> `ForwardForTraining` as one giant batch, completely ignoring `BatchSize`.

This is structurally redundant for gradient-based optimizers: the epoch loop's `CalculateGradient` + `UpdateSolution` is the actual training. The initial Train just produces one full-batch gradient step that the epoch loop will overwrite on its first iteration anyway. The comment at `GradientBasedOptimizerBase.cs:45-50` confirms the design:

> Gradient-based optimizers skip `model.Train()` during evaluation AFTER the initial training. They already update parameters via `UpdateSolution()` — re-training would overwrite gradient updates. The flag is set to true after the first `PrepareAndEvaluateSolution` call.

i.e. the *only* purpose of the initial Train is to "be a Train call before flipping the skip flag." Since `_m` / `_v` are deferred until the first `UpdateSolution` (`AdamOptimizer.cs:138-143`, comment cites #1221), there is no Adam-state dependency on the initial Train having happened. Skipping it entirely is functionally identical except for the OOM.

## Proposed fix

One-line change in `GradientBasedOptimizerBase.cs`:

```csharp
// Before
private bool _hasCompletedInitialTraining;

// After
private bool _hasCompletedInitialTraining = true;
```

This makes `SkipTrainingInEvaluation` return `true` from the very first call, so `PrepareAndEvaluateSolution` evaluates the (untrained) model directly without running `Model.Train(XTrain, YTrain)`. The epoch loop's mini-batched `CalculateGradient` + `UpdateSolution` is unchanged and does all the actual training.

The first `previousStepData` then reflects the untrained model's fitness, which is correct (it's the baseline we're improving from). Adam's convergence check (`AdamOptimizer.cs:182`) and `UpdateBestSolution` still work because they compare against `currentStepData` after the first epoch, which is now correctly the result of mini-batched training.

## Test coverage

The existing tape-retention regression suite (`tests/AiDotNet.Tests/IntegrationTests/NeuralNetworks/TransformerTrainPathReproIssue1227And1228Tests.cs`) is per-Train-call; this issue would need a new probe that exercises `Optimize()` on a large XTrain tensor and asserts peak heap stays under (say) 500 MB.

Suggested probe (sketch):

```csharp
[Fact]
public void AdamOptimize_LargeXTrain_RespectsBatchSize_NoOOM()
{
    int N = 50_000, ctx = 64, V = 256;
    var xTrain = new Tensor<float>([N, ctx]);
    var yTrain = new Tensor<float>([N, V]);
    // ... fill ...

    var arch = /* d=128, L=4, heads=4, ff=512, ctx=64, V=256 */;
    var model = new Transformer<float>(arch, lossFunction: new CategoricalCrossEntropyLoss<float>());
    var adamOpts = new AdamOptimizerOptions<float, Tensor<float>, Tensor<float>>
    {
        InitialLearningRate = 3e-4, MaxIterations = 1, BatchSize = 32,
    };
    var optimizer = new AdamOptimizer<float, Tensor<float>, Tensor<float>>(null, adamOpts);

    long heapBefore = GC.GetTotalMemory(forceFullCollection: true);
    var result = optimizer.Optimize(new OptimizationInputData<float, Tensor<float>, Tensor<float>>
    {
        XTrain = xTrain, YTrain = yTrain, /* ... */
    });
    long heapAfter = GC.GetTotalMemory(forceFullCollection: false);

    Assert.True(heapAfter - heapBefore < 500L * 1024 * 1024,
        $"Peak managed heap delta {(heapAfter - heapBefore) / 1024.0 / 1024.0:F1} MB exceeded 500 MB limit");
}
```

## Environment

- AiDotNet 0.190.0 (NuGet)
- AiDotNet.Tensors 0.75.6
- .NET 10, Windows 11
- Encountered while building a facade head-to-head harness for HRE substrate research; surfaces on any Transformer trained via `AiModelBuilder.BuildAsync` on a corpus larger than a few KB.

## Related

- #1227 (closed yesterday) — different issue, fixed cross-Train-call retention. This is a single-call OOM.
- #1228 (closed) — single-thread backward. Orthogonal.
- AdamOptimizerOptions.BatchSize has an existing accessor that the rest of the optimizer respects; this is the one path that doesn't.
